webgl2: Allow for better type union with `WebGLRenderingContext`

karhu opened the pull request:
This PR adjusts the typings for WebGL2 to allow for better merging of method declarations between `WebGLRenderingContext` and `WebGL2RenderingContext` (which extends the former). 

Part of the problem here is that `WebGLRenderingContext` typings are generated automatically (as part of `lib.dom.d.ts`), and over time, some of the method signatures of `WebGLRenderingContext` have changed between typescript versions, e.g.:

```
compressedTexImage2D(target: number, level: number, internalformat: number, width: number, height: number, border: number, data: Int8Array | Int16Array | Int32Array | Uint8Array | Uint16Array | Uint32Array | Uint8ClampedArray | Float32Array | Float64Array | DataView | null): void;
``` 
in `typescript@2.8` and
```
compressedTexImage2D(target: number, level: number, internalformat: number, width: number, height: number, border: number, data: ArrayBufferView): void;
```
in `typescript@2.0`.

Without these changes, when working with union types like in the following example, the typescript compiler fails to merge the (intended) to be same method declarations between the two contexts:

```
type WebGLContext = WebGLRenderingContext | WebGL2RenderingContext;

const gl: WebGLContext = ...;

gl.compressedTexImage2D(...); // compile error here, failing to resolve the method signature
```

karhu commented:
@nkemnitz could you have a look at this please?

nkemnitz commented:
Thanks for the PR, @karhu. I'll have a look at it later today or tomorrow. Meanwhile, I am going to add @mhegazy here because the `ArrayBufferView` inconsistencies are likely related to the recent changes in #24019. Those added `null` as acceptable `srcData` type for `bufferSubData`, `compressedTexImage2D` and `compressedTexSubImage2D`. See remarks at https://github.com/DefinitelyTyped/DefinitelyTyped/pull/24019#issuecomment-370910790 . Hopefully we can resolve the different interpretation of the spec here.

nkemnitz commented on types/webgl2/index.d.ts:
```diff
@@ -387,17 +392,30 @@ interface WebGL2RenderingContext extends WebGLRenderingContext {
 
     compressedTexImage3D(target: number, level: number, internalformat: number, width: number,
         height: number, depth: number, border: number, imageSize: number, offset: number): void;
+    compressedTexImage3D(target: number, level: number, internalformat: number, width: number,
```
That change shouldn't be required, I think. `compressedTexImage3D` was introduced in WebGL2, so it shouldn't clash with anything in lib.dom.d.ts right now.

If you want to keep the verbose version, please remove `null`. The [specification](https://www.khronos.org/registry/webgl/specs/latest/2.0/) states that `srcData` has to be an `ArrayBufferView` (which excludes `null`).

nkemnitz commented on types/webgl2/index.d.ts:
```diff
@@ -407,6 +425,12 @@ interface WebGL2RenderingContext extends WebGLRenderingContext {
     compressedTexSubImage3D(target: number, level: number, xoffset: number, yoffset: number,
         zoffset: number, width: number, height: number, depth: number,
         format: number, imageSize: number, offset: number): void;
+    compressedTexSubImage3D(target: number, level: number, xoffset: number, yoffset: number,
```
`compressedTexSubImage3D` was also introduced in WebGL2, so it shouldn't clash with anything in lib.dom.d.ts right now.

And here as well: If you want to keep the verbose version, please remove `null`. The [specification](https://www.khronos.org/registry/webgl/specs/latest/2.0/) states that `srcData` has to be an `ArrayBufferView` (which excludes `null`).

nkemnitz commented on types/webgl2/index.d.ts:
```diff
@@ -486,21 +510,23 @@ interface WebGL2RenderingContext extends WebGLRenderingContext {
         offset: number): void;
     readPixels(x: number, y: number, width: number, height: number, format: number, type: number,
-        dstData: ArrayBufferView, dstOffset: number): void;
+        dstData: Int8Array | Int16Array | Int32Array | Uint8Array | Uint16Array | Uint32Array | Uint8ClampedArray |
```
The overloaded 8 parameter version was introduced in WebGL2, and shouldn't clash with lib.dom.d.ts. If you want to keep the verbose version, please remove `null`. Specifications for WebGL1 and 2 differ here for some reason in how they treat `dstData`... not sure why / if intended.

nkemnitz commented on types/webgl2/index.d.ts:
```diff
@@ -486,21 +510,23 @@ interface WebGL2RenderingContext extends WebGLRenderingContext {
     // WebGL1:
     readPixels(x: number, y: number, width: number, height: number, format: number, type: number,
-        dstData: ArrayBufferView | null): void;
+        dstData: Int8Array | Int16Array | Int32Array | Uint8Array | Uint16Array | Uint32Array | Uint8ClampedArray |
```
Does this method also require a compatibility line for older Typescript versions?

karhu replied:
In the other cases I noticed the compatibility line was necessary when running `npm run lint webgl2`. Strangely enough it didn't trigger in this case. 

I would suggest still adding it, or WDYT?

karhu commented:
Thanks a lot for the fast review! Finally got around to address your feedback. 

Typos are fixed, and I introduced a compatibility line for `readPixels` but please see my inline note on that. Lastly I reverted the function signature changes in the cases where the function in question (or its overload) doesn't exist in WebGL1.
